enterprise/app/trends/heatmap.tsx

```diff
@@ -0,0 +1,422 @@
+import React from "react";
+import moment from "moment";
+
+import { stats } from "../../../proto/stats_ts_proto";
+import { ScaleBand, scaleBand } from "d3-scale";
+import { withResizeDetector } from "react-resize-detector";
+import { pinBottomLeftOffsetFromMouse, MouseCoords, Tooltip } from "../../../app/components/tooltip/tooltip";
+
+interface HeatmapProps {
+  heatmapData: stats.GetStatHeatmapResponse;
+  width: number;
+  height: number;
+  valueFormatter: (value: number) => string;
+  metricBucketName: string;
+  metricBucketFormatter: (value: number) => string;
+  selectionCallback?: (s?: HeatmapSelection) => void;
+}
+
+interface State {
+  selectionToRender?: [TooltipData, TooltipData];
+}
+
+type TooltipData = {
+  timestamp: number;
+  timestampBucketIndex: number;
+  metric: number;
+  metricBucketIndex: number;
+  value: number;
+};
+
+export type HeatmapSelection = {
+  dateRangeMicros: number[];
+  bucketRange: number[];
+  invocationsSelected: number;
+};
+
+const CHART_MARGINS = {
+  top: 1,
+  right: 1,
+  bottom: 39,
+  left: 89,
+};
+
+// This is a magic number that states there will only be one axis label for
+// every N pixels of rendered axis length (currently 100).
+const TICK_LABEL_SPACING_MAGIC_NUMBER = 100;
+const heatmapColorScale = [
+  "#d9d8ea",
+  "#cecee5",
+  "#c2c2df",
+  "#b6b5d8",
+  "#aaa8d0",
+  "#9e9bc9",
+  "#928ec3",
+  "#8782bc",
+  "#7c73b5",
+  "#7363ac",
+  "#6a51a3",
+  "#61409b",
+  "#583093",
+  "#501f8c",
+  "#471084",
+];
+
+type SelectionData = {
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+  selectionXStart: number;
+  selectionYStart: number;
+  selectionXEnd: number;
+  selectionYEnd: number;
+};
+
+class HeatmapComponentInternal extends React.Component<HeatmapProps, State> {
+  state: State;
+  svgRef: React.RefObject<SVGSVGElement>;
+  xScaleBand: ScaleBand<number>;
+  yScaleBand: ScaleBand<number>;
+  pendingClick?: [TooltipData, TooltipData];
+  selectedData?: [TooltipData, TooltipData];
+
+  constructor(props: HeatmapProps) {
+    super(props);
+    this.state = {};
+    this.xScaleBand = scaleBand();
+    this.yScaleBand = scaleBand();
+    this.svgRef = React.createRef();
+  }
```

Comment:
nit: might be slightly cleaner to initialize these in the top level of the component -- then the constructor can be removed

Reply:
Thanks, I had some meatier stuff in here before and forgot to clean this up.